cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -353,14 +353,10 @@ protected void startUp() throws Exception {
       // Need to wrap LinkageError. Otherwise, listeners of this Guava Service may not be called if the initialization
       // of the user program is missing dependencies (CDAP-2543)
       throw new Exception(e.getMessage(), e);
-    } catch (Throwable t) {
+    } catch (TransactionFailureException e) {
       cleanupTask.run();
```

Comment:
The cleanup has to run on all kinds of exception (notice in line 355, there is a `throw new Exception`).